### PR TITLE
Proper generalisation of well-founded relation in program logic

### PR DESCRIPTION
**Pull request by didriklundberg**

### CODE REVIEW

File: src/theory/program_logic/total_program_logicScript.sml [hunk 1]
```diff
@@ -141,56 +141,188 @@ QED
...
 val loop_fun_ind_spec =
- Q.SPEC `\TS s var l le invariant C1.
+ Q.SPEC `\TS s wf_rel var l le invariant C1.
           first_enc TS ==>
-	  total_loop_jgmt TS l le invariant C1 var ==>
+          WF wf_rel ==>
+	  total_loop_jgmt TS l le invariant C1 wf_rel var ==>
 	  t_jgmt TS l le (\s. invariant s /\ ~C1 s) post ==>
-	  (invariant s /\ TS.ctrl s = l) ==>
+	  (invariant s /\ TS.ctrl s = l /\ C1 s) ==> (* TODO: Why does this need "C1 s" as conjunct also? *)
```
Comment (arolle) [inline src/theory/program_logic/total_program_logicScript.sml:325]: I'm curious why earlier `C1 s` had not been needed for the loop invariant `loop_fun_ind_spec`.

**Review by arolle — DISMISSED** [review]
The failing tests seem to be unrelated to the changes.

File: src/theory/program_logic/total_program_logicScript.sml (hunk 1 above)
Comment (didriklundberg) [reply src/theory/program_logic/total_program_logicScript.sml]: Also curious. Tried removing it - no apparent harm done with minimal changes needed to compensate. No idea how it appeared there.

**Review by arolle — APPROVED** [review]
Thanks for resolving this.

**Pull request merged**